nequip/utils/gmm.py

```diff
@@ -0,0 +1,149 @@
+from typing import Optional
+
+import torch
+import numpy as np
+from sklearn import mixture
+
+
+@torch.jit.script
+def _compute_log_det_cholesky(
+    matrix_chol: torch.Tensor, covariance_type: str, n_features: int
+):
+    """Compute the log-det of the cholesky decomposition of matrices."""
+
+    assert covariance_type in ("full",)
+    n_components = matrix_chol.size(dim=0)
+    log_det_chol = torch.sum(
+        torch.log(matrix_chol.view(n_components, -1)[:, :: n_features + 1]), dim=1
+    )
+
+    return log_det_chol
+
+
+@torch.jit.script
+def _estimate_log_gaussian_prob(
+    X: torch.Tensor,
+    means: torch.Tensor,
+    precisions_chol: torch.Tensor,
+    covariance_type: str,
+):
+    """Estimate the log Gaussian probability."""
+
+    assert covariance_type in ("full",)
+    n_samples, n_features = X.size(dim=0), X.size(dim=1)
+    n_components = means.size(dim=0)
+
+    # det(precision_chol) = -0.5 * det(precision)
+    log_det = _compute_log_det_cholesky(precisions_chol, covariance_type, n_features)
+
+    # TODO: understand this
+    log_prob = torch.empty((n_samples, n_components))
+    for k, (mu, prec_chol) in enumerate(zip(means, precisions_chol)):
+        y = torch.mm(X, prec_chol) - torch.mm(mu.unsqueeze(0), prec_chol)
+        log_prob[:, k] = torch.sum(torch.square(y), dim=1)
+    # X = X.unsqueeze(0)
+    # means = means.unsqueeze(2)
+    # print(f"X shape: {X.size()}")
+    # print(f"means shape: {means.size()}")
+    # log_prob = torch.sum(
+    #     torch.square(
+    #         torch.matmul(X, precisions_chol.transpose(1, 2))
+    #         - torch.matmul(means, precisions_chol.transpose(1, 2))
+    #     ),
+    #     dim=1,
+    # )
+
+    # Since we are using the precision of the Cholesky decomposition,
+    # `-0.5 * log_det` becomes `+ log_det`
+    return -0.5 * (n_features * torch.log(torch.tensor(2 * np.pi)) + log_prob) + log_det
+
+
+class GaussianMixture(torch.nn.Module):
+    """Calculate NLL of samples under a Gaussian Mixture Model (GMM).
+
+    Supports fitting the GMM outside of PyTorch using `sklearn`.
+    """
+
+    covariance_type: str
+    n_components: int
+    n_features: int
+    seed: int
+
+    def __init__(
+        self,
+        n_components: Optional[int],
+        seed: Optional[int],
+        n_features: int,
+        covariance_type: str = "full",
+    ):
+        super(GaussianMixture, self).__init__()
+        assert covariance_type in ("full",)
+        self.covariance_type = covariance_type
+        self.n_components = n_components
+        self.n_features = n_features
+        self.seed = seed
+
+        self.register_buffer("means", torch.zeros(n_components, n_features))
```

Comment:
can set these to `torch.Tensor()` (the empty tensor) by default instead, then can check if fitted in froward